d2/src/main/java/com/linkedin/d2/balancer/clusterfailout/FailedoutClusterManager.java

```diff
@@ -155,16 +165,23 @@ private void addClusterWatches(@Nonnull Set<String> clustersToWatch, @Nonnull Fa
     }
     for (final String cluster : clustersToWatch)
     {
-      _clusterListenerCallbacks.computeIfAbsent(cluster, clusterName -> {
-        LoadBalancerStateListenerCallback listenerCallback = (type, name) -> {
+      _peerWatches.computeIfAbsent(cluster, clusterName ->
+      {
+        boolean watchExistsBeforeFailout = _loadBalancerState.isListeningToCluster(clusterName);
+        PeerWatchState peerWatchState = new PeerWatchState(watchExistsBeforeFailout);
+
+        LoadBalancerStateListenerCallback listenerCallback = (type, name) ->
```

Comment:
Same here, do we want to use a FutureCallback, and provide a timeout value? We can probably use the same timeout value as `SimpleLoadBalancer._timeout`

Reply:
Looking at the logic, I think we don't need to block/wait for the call to finish. It will be a best effort clean-up.